roles/doom/files/.doom.d/config.org
#+property: header-args:elisp :tangle config.el
#+startup: fold

#+TITLE: Doom Emacs config
#+author: Luke Piotrak
#+email: luke.piotrak

My Doom Emacs configuration. Most snippets have been copied from various places, including:

 - [[https://github.com/jethrokuan/dots/blob/master/.doom.d/config.el][jethrokuan/dots]]
 - [[https://tecosaur.github.io/emacs-config/config.html][tecosaur's emacs config]]

If you'd like to roll your own literate config I highly recommend the [[https://leanpub.com/lit-config/read][excellent guide]] by [[https://zzamboni.org/post/][Diego Zamboni]].

* Table of contents :TOC_4:noexport:
- [[#useful-resources][Useful resources]]
- [[#a-note-on-this-files-structure][A note on this file's structure]]
- [[#literate-config-setup][Literate config setup]]
- [[#general][General]]
- [[#custom-keymap][Custom Keymap]]
- [[#appearance][Appearance]]
- [[#org][Org]]
  - [[#general-org-mode][General Org-mode]]
  - [[#org-mode-appearance][Org-mode appearance]]
  - [[#org-agenda-and-gtd][Org Agenda and GTD]]
    - [[#org-capture-setup][Org Capture setup]]
    - [[#tags][Tags]]
    - [[#custom-functions-for-gtd][Custom functions for GTD]]
  - [[#org-roam][Org Roam]]
  - [[#org-roam-server][Org Roam server]]
    - [[#org-roam-capture-templates][Org roam capture templates]]
  - [[#org-pomodoro][Org Pomodoro]]
  - [[#org-ref][Org-ref]]
- [[#auctex][AUCTeX]]
  - [[#appearance-1][Appearance]]
- [[#file-templates][File templates]]
  - [[#configel][config.el]]
  - [[#initel][init.el]]
  - [[#packagesel][packages.el]]

* Useful resources

Before we begin, here are a list of resources I have found useful when configuring and using Emacs:
- [[https://leanpub.com/lit-config/read#leanpub-auto-tips-and-tricks][Read Literate Configuration | Leanpub]]
- [[https://blog.jethro.dev/posts/capturing_inbox/][Org-mode Workflow Part 1: Capturing in the Inbox · Jethro Kuan]]
- [[https://narra.co.za/doom-tips-and-howtos/][Doom Tips and Howtos - Narra's Place]] 

* A note on this file's structure
This file generates all three Doom configuration files (namely, [[file:init.el][init.el]], [[file:config.el][config.el]] and [[file:packages.el][packages.el]]). This makes it possible to group snippets from all three files pertaining to a logical part of the Emacs ecosystem to live under one heading. I find this more logical, and easier to manage, rather than the default Doom Emacs split.

The templates for all three files can be found at the bottom of this config.

* Literate config setup
:PROPERTIES:
:header-args:elisp: :noweb-ref preamble
:END:

First of all, I want the configs to be nice and comfy to edit, so let's do that.
The hook below automatically generates the configs once this file is saved.

#+begin_src elisp
(add-hook 'org-mode-hook
          (lambda () (add-hook 'after-save-hook #'org-babel-tangle
                               :append :local)))
#+end_src

Pdf tools is needed for generating the files. It is also used in org-noter
#+begin_src elisp
(pdf-tools-install)
#+end_src
* General
:PROPERTIES:
:header-args:elisp: :noweb-ref preamble
:END:

Enable fuzzy searching in search buffers
#+begin_src elisp :noweb-ref ivy-init.el
(ivy +fuzzy)
#+end_src

Set autosaving
#+begin_src elisp
(setq auto-save-visited-mode t)
#+end_src

line numbers
#+begin_src elisp
(setq display-line-numbers-type nil)
#+end_src

Make avy text searches span all windows
#+begin_src elisp
(setq avy-all-windows t)
#+end_src

Bookmarks get saved to with every modification
#+begin_src elisp
(setq bookmark-save-flag 1)
#+end_src

Always show hints for ace-window so it is easy to see which window we can jump to
#+begin_src elisp
(define-globalized-minor-mode ace-window-display-mode-global ace-window-display-mode
  (lambda () (ace-window-display-mode 1)))
(ace-window-display-mode-global 1)
#+end_src

Centaur tabs
#+begin_src elisp
(after! centaur-tabs
  (centaur-tabs-mode 1)
  (setq centaur-tabs-height 36
        centaur-tabs-set-icons t
        centaur-tabs-modified-marker "o"
        centaur-tabs-close-button "×"
        centaur-tabs-set-bar 'above
        centaur-tabs-gray-out-icons 'buffer)
  )
#+end_src

Vue development
#+begin_src elisp
(add-hook 'vue-mode-hook #'lsp!)
#+end_src
* Custom Keymap
:PROPERTIES:
:header-args:elisp: :noweb-ref preamble
:END:

Delete current file and buffer keymap
#+begin_src elisp
(defun delete-file-and-buffer ()
  "Kill the current buffer and deletes the file it is visiting."
  (interactive)
  (let ((filename (buffer-file-name)))
    (if filename
        (if (y-or-n-p (concat "Do you really want to delete file " filename " ?"))
            (progn
              (delete-file filename)
              (message "Deleted file %s." filename)
              (kill-buffer)))
      (message "Not a file visiting buffer!"))))

(map! :leader
      :prefix "b"
      :desc "delete-file-and-buffer" "D" #'delete-file-and-buffer)
#+end_src

Make evil-mode up/down operate in screen lines instead of logical lines
#+begin_src elisp
(define-key evil-motion-state-map "j" 'evil-next-visual-line)
(define-key evil-motion-state-map "k" 'evil-previous-visual-line)
;; Also in visual mode
(define-key evil-visual-state-map "j" 'evil-next-visual-line)
(define-key evil-visual-state-map "k" 'evil-previous-visual-line)
#+end_src

Replace with register
#+begin_src elisp :noweb-ref packages
(package! evil-replace-with-register
  :recipe (:host github :repo "Dewdrops/evil-ReplaceWithRegister"))
#+end_src

#+begin_src elisp
(require 'evil-replace-with-register)
(setq evil-replace-with-register-key (kbd "gr"))
(evil-replace-with-register-install)
#+end_src

Undo / redo action
#+begin_src elisp
(winner-mode +1)
(map! :map winner-mode-map
      "<C-M-right>" #'winner-redo
      "<C-M-left>" #'winner-undo)
#+end_src

Undo / redo movement
#+begin_src elisp
(map! :leader
      "<M-left>" #'avy-pop-mark)
#+end_src
* Appearance
:PROPERTIES:
:header-args:elisp: :noweb-ref preamble
:END:

Let's go ahead and set the beautiful Hack font with all the bells and whistles. Can be found [[https://github.com/pyrho/hack-font-ligature-nerd-font/blob/master/font/Hack%20Regular%20Nerd%20Font%20Complete%20Mono.ttf/][here]] (thanks [[https://github.com/pyrho][pyrho]]).
#+begin_src elisp
(setq doom-font (font-spec :family "Hack Nerd Font Mono" :size 15 :height 1.0)
      ;; doom-variable-pitch-font (font-spec :family "Newsreader" :height 1.0 :size 17)
      )
#+end_src

Ligatures. I like ligatures
#+begin_src elisp :noweb-ref ligatures-init.el
(ligatures)
#+end_src

# Set the theme:
# First define the external package in packages.el
# #+begin_src elisp :noweb-ref packages
# (package! spacemacs-theme)
# #+end_src
# And set in config.el
# #+begin_src elisp
# (setq doom-theme 'spacemacs-light)
# #+end_src

# Disable solaire mode and line highlighting
# #+begin_src elisp
#   (setq solaire-global-mode -1)
#   (setq global-hl-line-mode -1)
# #+end_src

#+RESULTS:
: -1

* Org
:PROPERTIES:
:header-args:elisp: :noweb-ref preamble
:END:

I have recently begun to use org-mode quite heavily. Mainly for tracking todos and note-taking using org-roam.

** General Org-mode

Enable org-mode module with extras in init.el
#+begin_src elisp :noweb-ref org-init.el
(org +dragndrop +roam +pomodoro +pretty +noter :tools pdf)
                                        ;organize your plain life in plain text
#+end_src

Define where org files can be found
#+begin_src elisp

(require 'find-lisp)
(setq org-directory "~/Dropbox/org/"
      jethro/org-agenda-directory (concat org-directory "gtd/")
      org-agenda-files (find-lisp-find-files jethro/org-agenda-directory "\.org$")
      org-startup-folded 'overview)
#+end_src

Todo keywords
#+begin_src elisp 
(setq org-todo-keywords
      '((sequence "TODO(t)" "NEXT(n)" "HOLD(h)" "|" "DONE(d)")
        ))
#+end_src

Custom column display
#+begin_src elisp
  (setq org-columns-default-format "%40ITEM(Task) %Effort(EE){:} %CLOCKSUM(Time Spent) %SCHEDULED(Scheduled) %DEADLINE(Deadline)")
#+end_src
# ** Org-mode appearance

# Set mixed pitch mode to be available in org mode.
# #+begin_src elisp
# (use-package! mixed-pitch
#   :hook (org-mode . mixed-pitch-mode)
#   :config
#   (setq mixed-pitch-set-height t)
#   (set-face-attribute 'variable-pitch nil :height 1.3))

# (add-hook 'org-mode-hook :append :local
#           (lambda ()
#             (mixed-pitch-mode 1)
#             (hl-line-mode -1)
#             (set-left-margin 4)
#             (set-right-margin 4)
#             (set-window-buffer nil (current-buffer))))

# (setq line-spacing 0.1
#       header-line-format " "
#       )

# (setq org-startup-indented t
#       org-superstar-headline-bullets-list '(" ") ;; no bullets, needs org-bullets package
#       org-ellipsis "  " ;; folding symbol
#       org-pretty-entities t
#       org-hide-emphasis-markers t
#       ;; show actually italicized text instead of /italicized text/
#       org-agenda-block-separator ""
#       org-fontify-whole-heading-line t
#       org-fontify-done-headline t
#       org-fontify-quote-and-verse-blocks t)
# #+end_src

# Styles inspired by [[https://lepisma.xyz/2017/10/28/ricing-org-mode/][Ricing up Org Mode]] and adapted for doom
# #+begin_src elisp
# (setq-local
#  bg-white           "#fbf8ef"
#  bg-light           "#222425"
#  bg-dark            "#1c1e1f"
#  bg-darker          "#1c1c1c"
#  fg-white           "#ffffff"
#  shade-white        "#efeae9"
#  fg-light           "#655370"
#  dark-cyan          "#008b8b"
#  region-dark        "#2d2e2e"
#  region             "#39393d"
#  slate              "#8FA1B3"
#  keyword            "#f92672"
#  comment            "#525254"
#  builtin            "#fd971f"
#  purple             "#9c91e4"
#  doc                "#727280"
#  type               "#66d9ef"
#  string             "#b6e63e"
#  gray-dark          "#999"
#  gray               "#bbb"
#  sans-font          "Hack Nerd Font Mono"
#  serif-font         "Newsreader"
#  et-font            "Newsreader"
#  sans-mono-font     "Hack Nerd Font Mono"
#  serif-mono-font    "Verily Serif Mono")

# (custom-theme-set-faces
#  'user
#  `(variable-pitch

#    ( (t (:family ,et-font
#          :foreground ,bg-dark
#          :background ,bg-white
#          :height 1.7) ) ))
#  `(org-document-title

#    ( (t (:inherit nil
#          :family ,et-font
#          :height 1.8
#          :foreground ,bg-dark
#          :underline nil) ) ))
#  `(org-document-info

#    ( (t (:height 1.2
#          :slant italic) ) ))
#  `(org-level-1

#    ( (t (:inherit nil
#          :family ,et-font
#          :height 1.6
#          :weight normal
#          :slant normal
#          :foreground ,bg-dark) ) ))
#  `(org-level-2

#    ( (t (:inherit nil
#          :family ,et-font
#          :weight normal
#          :height 1.3
#          :slant italic
#          :foreground ,bg-dark) ) ))
#  `(org-level-3

#    ( (t (:inherit nil
#          :family ,et-font
#          :weight normal
#          :slant italic
#          :height 1.2
#          :foreground ,bg-dark) ) ))
#  `(org-level-4

#    ( (t (:inherit nil
#          :family ,et-font
#          :weight normal
#          :slant italic
#          :height 1.1
#          :foreground ,bg-dark) ) ))
#  `(org-level-5
#    ( (t (:inherit variable-pitch
#          :weight bold
#          :height 1.1
#          :foreground ,slate) ) )
#    )
#  `(org-level-6
#    ( (t (:inherit variable-pitch
#          :weight bold
#          :height 1.1
#          :foreground ,slate) ) )
#    )
#  `(org-level-7
#    ( (t (:inherit variable-pitch
#          :weight bold
#          :height 1.1
#          :foreground ,slate) ) )
#    )
#  `(org-level-8
#    ( (t (:inherit variable-pitch
#          :weight bold
#          :height 1.1
#          :foreground ,slate) ) )
#    )
#  `(org-headline-done

#    ( (t (:family ,et-font
#          :strike-through t) ) ))
#  `(org-quote
#    ( (t (:background ,bg-light) ) )
#    )
#  `(org-block

#    ( (t (:background nil
#          :foreground ,bg-dark) ) ))
#  `(org-block-begin-line

#    ( (t (:background nil
#          :height 0.8
#          :family ,sans-mono-font
#          :foreground ,bg-dark) ) ))
#  `(org-block-end-line

#    ( (t (:background nil
#          :height 0.8
#          :family ,sans-mono-font
#          :foreground ,bg-dark) ) ))
#  `(org-document-info-keyword

#    ( (t (:height 0.8
#          :foreground ,gray) ) ))
#  `(org-link

#    ( (t (:foreground ,bg-dark) ) ))
#  `(org-special-keyword

#    ( (t (:family ,sans-mono-font
#          :height 0.8) ) ))
#  `(org-todo
#    ( (t (:inherit variable-pitch :height 0.8) ) )
#    )
#  `(org-done
#    ( (t (:inherit variable-pitch
#          :height 0.8
#          :strike-through t
#          ) ) )
#    )
#  `(org-agenda-current-time
#    ( (t (:foreground ,slate) ) )
#    )
#  `(org-hide
#    ( (t (:foreground ,bg-white) ) ))
#  `(org-indent
#    ( (t (:inherit org-hide) ) )
#    )
#  `(org-time-grid
#    ( (t (:foreground ,comment) ) )
#    )
#  `(org-warning
#    ( (t (:foreground ,builtin) ) )
#    )
#  `(org-date
#    ( (t (:family ,sans-mono-font
#          :height 0.8) ) ))
#  `(org-agenda-structure
#    ( (t (:height 1.3
#          :foreground ,doc
#          :weight normal
#          :inherit variable-pitch) ) )
#    )
#  `(org-agenda-date

#    ( (t (:inherit variable-pitch
#          :height 1.1) ) ))
#  `(org-agenda-date-today
#    ( (t (:height 1.5
#          :foreground ,dark-cyan
#          :inherit variable-pitch) ) )
#    )
#  `(org-agenda-date-weekend
#    ( (t (:inherit org-agenda-date) ) )
#    )
#  `(org-scheduled
#    ( (t (:foreground ,gray) ) )
#    )
#  `(org-upcoming-deadline
#    ( (t (:foreground ,keyword) ) )
#    )
#  `(org-scheduled-today
#    ( (t (:foreground ,fg-white) ) )
#    )
#  `(org-scheduled-previously
#    ( (t (:foreground ,slate) ) )
#    )
#  `(org-agenda-done
#    ( (t (:strike-through t) ) ))
#  `(org-ellipsis

#    ( (t (:underline nil
#          :foreground ,comment) ) ))
#  `(org-tag

#    ( (t (:foreground ,doc) ) ))
#  `(org-table

#    ( (t (:family ,serif-mono-font
#          :height 0.9
#          :background ,bg-white) ) ))
#  `(org-code

#    ( (t (:inherit
#          :family ,serif-mono-font
#          :foreground ,comment
#          :height 0.7) ) ))
#  `(mode-line
#    ( (t (:background ,bg-white) ) ))
#  )
# #+end_src

# #+RESULTS:

# ** Org Agenda and GTD

# This workflow (loosely based on David Allen's GTD methodology) has mostly been taken from Jethro Kuan's [[https://blog.jethro.dev/posts/capturing_inbox/][great guide]] and modified it to preference.
# In short, any ideas, links, todos etc are org-captured to an inbox file. These later get processed in bulk (tagged, estimated efforts etc.) and refiled to the appropriate files (next, someday, projects etc.). A customized Agenda view is crucial in making this work.

# First of all, let's set up a shortcut for org-capture and define the templates to use. This is the entrypoint for the GTD system. The org-protocol-capture template is used mostly with [[file:../misc/org-capture-bookmarklet][this]] bookmarklet in firefox to easily push links from the browser to the inbox.
# *** Org Capture setup
# #+begin_src elisp
# (map! :leader
#       "C" #'org-capture)

# (setq org-capture-templates
#       `(("i" "Inbox" entry (file ,(concat jethro/org-agenda-directory "inbox.org"))
#          ,(concat "* TODO %?\n"
#                   "/Entered on/ %u \n"
#                   "What is the first step for this task?"))
#         ("c" "org-protocol-capture" entry (file ,(concat jethro/org-agenda-directory "inbox.org"))
#          "* TODO [[%:link][%:description]]\n\n %i"
#          :immediate-finish t)))

# (defun jethro/org-inbox-capture ()
#   (interactive)
#   "Capture a task in agenda mode."
#   (org-capture nil "i"))
# #+end_src

# *** Tags

# GTD items are annotated with tags. These represent represent:
# - areas (of responsibility) prefixed with @. *A sphere of activity with a standard to be maintained over time.*
# - Interests (no prefix)
# #+begin_src elisp

# (setq org-complete-tags-always-offer-all-agenda-tags t)
# (setq org-tag-alist '(("@social" . ?s)
#                       ("@work" . ?w)
#                       ("@music" . ?m)
#                       ("@fitness&health" . ?f)
#                       ("@learning" . ?l)
#                       ))
# #+end_src

# *** Custom functions for GTD

# Keybindings
# #+begin_src elisp
# (map! :map org-agenda-mode-map
#       "i" #'org-agenda-clock-in
#       "r" #'jethro/org-process-inbox
#       "R" #'org-agenda-refile
#       "c" #'jethro/org-inbox-capture)
# #+end_src

# Enabling refiling headings to other org agenda files
# #+begin_src elisp
# (setq org-refile-use-outline-path 'file
#       org-outline-path-complete-in-steps nil)
# (setq org-refile-allow-creating-parent-nodes 'confirm
#       org-refile-targets '((org-agenda-files . (:level . 1))))
# #+end_src

# Define functions to be called on marked agenda items
# #+begin_src elisp
# (defvar jethro/org-agenda-bulk-process-key ?f
#   "Default key for bulk processing inbox items.")
# (setq org-agenda-bulk-custom-functions `((,jethro/org-agenda-bulk-process-key jethro/org-agenda-process-inbox-item)))
# #+end_src

# Functions for processing the gtd inbox from the agenda view. The bulk inbox processing function needed a bit of hacking compared to jethro's original to actually work in my config. The original org-agenda-bulk-mark-regexp function didn't work as it tried to match text from the org *txt* category. Instead we want to match the *org-category* category. The modified function was taken from [[https://emacs.stackexchange.com/questions/59657/how-to-bulk-mark-agenda-items-based-on-file-name][here]]. This still had a problem though. Since one of the headings in the custom org agenda view is named *Inbox* The first regexp search matched the heading and the get-text-property call failed because Agenda headings don't have and *org-category* property, making the whole while condition fail. I fixed this by toggling regexp case sensitivity for the duration of the regexp. An ugly fix but it'll do until I have time to rewrite.
# #+begin_src elisp
# (defun jethro/org-process-inbox ()
#   "Called in org-agenda-mode, processes all inbox items."
#   (interactive)
#   (setq case-fold-search nil)
#   (custom/org-agenda-bulk-mark-regexp-category "inbox")
#   (setq case-fold-search t)
#   (jethro/bulk-process-entries))

# (defun custom/org-agenda-bulk-mark-regexp-category (regexp)
#   "Mark entries whose category matches REGEXP for future agenda bulk action."
#   (interactive "sMark entries with category matching regexp: ")
#   (let ((entries-marked 0) txt-at-point)
#     (save-excursion
#       (goto-char (point-min))
#       (goto-char (next-single-property-change (point) 'org-hd-marker))
#       (while (and (re-search-forward regexp nil t)
#                   (setq category-at-point
#                         (get-text-property (match-beginning 0) 'org-category)))
#         (if (get-char-property (point) 'invisible)
#             (beginning-of-line 2)
#           (when (string-match-p regexp category-at-point)
#             (setq entries-marked (1+ entries-marked))
#             (call-interactively 'org-agenda-bulk-mark)))))
#     (unless entries-marked
#       (message "No entry matching this regexp."))))

# (defun jethro/org-agenda-process-inbox-item ()
#   "Process a single item in the org-agenda."
#   (org-with-wide-buffer
#    (org-agenda-set-tags)
#    (org-agenda-priority)
#    (call-interactively 'jethro/my-org-agenda-set-effort)
#    (org-agenda-refile nil nil t)))

# (defun jethro/bulk-process-entries ()
#   (if (not (null org-agenda-bulk-marked-entries))
#       (let ((entries (reverse org-agenda-bulk-marked-entries))
#             (processed 0)
#             (skipped 0))
#         (dolist (e entries)
#           (let ((pos (text-property-any (point-min) (point-max) 'org-hd-marker e)))
#             (if (not pos)
#                 (progn (message "Skipping removed entry at %s" e)
#                        (cl-incf skipped))
#               (goto-char pos)
#               (let (org-loop-over-headlines-in-active-region) (funcall 'jethro/org-agenda-process-inbox-item))
#               ;; `post-command-hook' is not run yet.  We make sure any
#               ;; pending log note is processed.
#               (when (or (memq 'org-add-log-note (default-value 'post-command-hook))
#                         (memq 'org-add-log-note post-command-hook))
#                 (org-add-log-note))
#               (cl-incf processed))))
#         (org-agenda-redo)
#         (unless org-agenda-persistent-marks (org-agenda-bulk-unmark-all))
#         (message "Acted on %d entries%s%s"
#                  processed
#                  (if (= skipped 0)
#                      ""
#                    (format ", skipped %d (disappeared before their turn)"
#                            skipped))
#                  (if (not org-agenda-persistent-marks) "" " (kept marked)")))))
# #+end_src

# A function for interactively setting task effort
# #+begin_src elisp
# (defvar jethro/org-current-effort "1:00"
#   "Current effort for agenda items.")

# (defun jethro/my-org-agenda-set-effort (effort)
#   "Set the effort property for the current headline."
#   (interactive
#    (list (read-string (format "Effort [%s]: " jethro/org-current-effort) nil nil jethro/org-current-effort)))
#   (setq jethro/org-current-effort effort)
#   (org-agenda-check-no-diary)
#   (let* ((hdmarker (or (org-get-at-bol 'org-hd-marker)
#                        (org-agenda-error)))
#          (buffer (marker-buffer hdmarker))
#          (pos (marker-position hdmarker))
#          (inhibit-read-only t)
#          newhead)
#     (org-with-remote-undo buffer
#       (with-current-buffer buffer
#         (widen)
#         (goto-char pos)
#         (org-show-context 'agenda)
#         (funcall-interactively 'org-set-effort nil jethro/org-current-effort)
#         (end-of-line 1)
#         (setq newhead (org-get-heading)))
#       (org-agenda-change-all-lines newhead hdmarker))))
# #+end_src

# Change todo state if task clocked
# #+begin_src elisp
# (defun jethro/set-todo-state-next ()
#   "Visit each parent task and change NEXT states to TODO"
#   (org-todo "NEXT"))

# (add-hook 'org-clock-in-hook 'jethro/set-todo-state-next 'append)
# #+end_src

# Custom agenda view
# #+begin_src elisp
# (use-package! org-agenda
#   :init
#   (map! :leader
#         :prefix "n"
#         :desc "org-agenda" "a" #'jethro/switch-to-agenda)
#   (setq org-agenda-block-separator nil
#         org-agenda-start-with-log-mode t)
#   (defun jethro/switch-to-agenda ()
#     (interactive)
#     (org-agenda nil " "))
#   :config
#   (defun jethro/is-project-p ()
#     "Any task with a todo keyword subtask"
#     (save-restriction
#       (widen)
#       (let ((has-subtask)
#             (subtree-end (save-excursion (org-end-of-subtree t)))
#             (is-a-task (member (nth 2 (org-heading-components)) org-todo-keywords-1)))
#         (save-excursion
#           (forward-line 1)
#           (while (and (not has-subtask)
#                       (< (point) subtree-end)
#                       (re-search-forward "^\*+ " subtree-end t))
#             (when (member (org-get-todo-state) org-todo-keywords-1)
#               (setq has-subtask t))))
#         (and is-a-task has-subtask))))

#   (defun jethro/skip-projects ()
#     "Skip trees that are projects"
#     (save-restriction
#       (widen)
#       (let ((next-headline (save-excursion (or (outline-next-heading) (point-max)))))
#         (cond
#          ((jethro/is-project-p)
#           next-headline)
#          (t
#           nil)))))
# #+end_src

# Agenda prefixes to show the effort estimate
# #+begin_src elisp
#   (setq org-agenda-prefix-format '((agenda . " %i %-12:c%?-12t%-6e% s")
#                                    (todo . " %i %-12:c %-6e")
#                                    (tags . " %i %-12:c")
#                                    (search . " %i %-12:c")))
# #+end_src

# #+begin_src elisp
#   (setq org-agenda-custom-commands `((" " "Agenda"
#                                       ((agenda ""
#                                                ((org-agenda-span 'week)
#                                                 (org-deadline-warning-days 14)))
#                                        (todo "TODO"
#                                              ((org-agenda-overriding-header "Inbox")
#                                               (org-agenda-files '(,(concat jethro/org-agenda-directory "inbox.org")))))
#                                        (todo "TODO"
#                                              ((org-agenda-overriding-header "Emails")
#                                               (org-agenda-files '(,(concat jethro/org-agenda-directory "emails.org")))))
#                                        (todo "NEXT"
#                                              ((org-agenda-overriding-header "In Progress")
#                                               (org-agenda-files '(,(concat jethro/org-agenda-directory "projects.org") ,(concat jethro/org-agenda-directory "next.org")))))
#                                        (todo "TODO"
#                                              ((org-agenda-overriding-header "Active Projects")
#                                               (org-agenda-skip-function #'jethro/skip-projects)
#                                               (org-agenda-files '(,(concat jethro/org-agenda-directory "projects.org")))))
#                                        (todo "TODO"
#                                              ((org-agenda-overriding-header "One-off Tasks")
#                                               (org-agenda-files '(,(concat jethro/org-agenda-directory "next.org")))
#                                               (org-agenda-skip-function '(org-agenda-skip-entry-if 'deadline 'scheduled))))
#                                        (todo "TODO"
#                                              ((org-agenda-overriding-header "Habits")
#                                               (org-agenda-files '(,(concat jethro/org-agenda-directory "habits.org")))
#                                               (org-agenda-skip-function '(org-agenda-skip-entry-if 'deadline 'scheduled))))

#                                        )))))
# #+end_src
# ** Org Roam
# #+begin_src elisp
# (setq org-roam-directory (concat org-directory "roam/") )
# (require 'org-roam-protocol)
# (map! :leader
#       :prefix "n"
#       :desc "org-roam" "l" #'org-roam
#       :desc "org-roam-insert" "i" #'org-roam-insert
#       :desc "org-roam-switch-to-buffer" "b" #'org-roam-switch-to-buffer
#       :desc "org-roam-find-file" "f" #'org-roam-find-file
#       :desc "org-roam-show-graph" "g" #'org-roam-show-graph
#       :desc "org-roam-capture" "c" #'org-roam-capture
#       :desc "org-roam-deactivate-buffer" "d" #'org-roam-buffer-toggle-display)

# (setq org-roam-tag-sources '(prop last-directory))
# #+end_src
# ** Org Roam server
# #+begin_src elisp :noweb-ref packages
# (package! org-roam-server :recipe (:host github :repo "org-roam/org-roam-server" :files ("*")))
# #+end_src

# #+begin_src elisp
# (use-package! org-roam-server
#   :after ( org-roam server )
#   :config
#   (setq org-roam-server-host "127.0.0.1"
#         org-roam-server-port 8080
#         org-roam-server-export-inline-images t
#         org-roam-server-authenticate nil
#         )
#   (defun org-roam-server-open ()
#     "Ensure the server is active, then open the roam graph."
#     (interactive)
#     (org-roam-server-mode 1)
#     (browse-url-xdg-open (format "http://localhost:%d" org-roam-server-port))))
# #+end_src

# *** Org roam capture templates

# Roam supports capture functionality. We need to escape these by prepending a ',' before headings.
# #+begin_src elisp :noweb-ref roam-templates

# ;; TODO template the main string
# (setq org-roam-capture-templates
#         '(("r" "read" plain (function org-roam--capture-get-point)
#            "%?"
#            :file-name "src/read/${slug}"
#            :head "#+title: ${title}\n
# #+roam_alias: \n
# #+roam_tags: \n
# ,* ${title}\n
# - source :: ${ref}\n
# - links :: "
#            :unnarrowed t)
#           ("w" "web" plain (function org-roam--capture-get-point)
#            "%?"
#            :file-name "src/web/${slug}"

roles/doom/files/.doom.d/config.el
(add-hook 'LaTeX-mode-hook (lambda ()
                             (TeX-fold-mode 1)))

;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; DO NOT EDIT THIS FILE DIRECTLY
;; This is a file generated from a literate programing source file located at
;; https://gitlab.com/zzamboni/dot-doom/-/blob/master/doom.org
;; You should make any changes there and regenerate it from Emacs org-mode
;; using org-babel-tangle (C-c C-v t)

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!

;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
;; (setq user-full-name "John Doe"
;;      user-mail-address "john@doe.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
;; (setq doom-theme 'doom-one)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
;; (setq org-directory "~/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
;; (setq display-line-numbers-type t)

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.


(add-hook 'org-mode-hook
          (lambda () (add-hook 'after-save-hook #'org-babel-tangle
                               :append :local)))
(pdf-tools-install)
(setq auto-save-visited-mode t)
(setq display-line-numbers-type nil)
(setq avy-all-windows t)
(setq bookmark-save-flag 1)
(define-globalized-minor-mode ace-window-display-mode-global ace-window-display-mode
  (lambda () (ace-window-display-mode 1)))
(ace-window-display-mode-global 1)
(after! centaur-tabs
  (centaur-tabs-mode 1)
  (setq centaur-tabs-height 36
        centaur-tabs-set-icons t
        centaur-tabs-modified-marker "o"
        centaur-tabs-close-button "×"
        centaur-tabs-set-bar 'above
        centaur-tabs-gray-out-icons 'buffer)
  )
(add-hook 'vue-mode-hook #'lsp!)
(defun delete-file-and-buffer ()
  "Kill the current buffer and deletes the file it is visiting."
  (interactive)
  (let ((filename (buffer-file-name)))
    (if filename
        (if (y-or-n-p (concat "Do you really want to delete file " filename " ?"))
            (progn
              (delete-file filename)
              (message "Deleted file %s." filename)
              (kill-buffer)))
      (message "Not a file visiting buffer!"))))

(map! :leader
      :prefix "b"
      :desc "delete-file-and-buffer" "D" #'delete-file-and-buffer)
(define-key evil-motion-state-map "j" 'evil-next-visual-line)
(define-key evil-motion-state-map "k" 'evil-previous-visual-line)
;; Also in visual mode
(define-key evil-visual-state-map "j" 'evil-next-visual-line)
(define-key evil-visual-state-map "k" 'evil-previous-visual-line)
(require 'evil-replace-with-register)
(setq evil-replace-with-register-key (kbd "gr"))
(evil-replace-with-register-install)
(winner-mode +1)
(map! :map winner-mode-map
      "<C-M-right>" #'winner-redo
      "<C-M-left>" #'winner-undo)
(map! :leader
      "<M-left>" #'avy-pop-mark)
(setq doom-font (font-spec :family "Hack Nerd Font Mono" :size 15 :height 1.0)
      ;; doom-variable-pitch-font (font-spec :family "Newsreader" :height 1.0 :size 17)
      )

(require 'find-lisp)
(setq org-directory "~/Dropbox/org/"
      jethro/org-agenda-directory (concat org-directory "gtd/")
      org-agenda-files (find-lisp-find-files jethro/org-agenda-directory "\.org$")
      org-startup-folded 'overview)
(setq org-todo-keywords
      '((sequence "TODO(t)" "NEXT(n)" "HOLD(h)" "|" "DONE(d)")
        ))
  (setq org-columns-default-format "%40ITEM(Task) %Effort(EE){:} %CLOCKSUM(Time Spent) %SCHEDULED(Scheduled) %DEADLINE(Deadline)")
(require 'org-ref)
(setq reftex-default-bibliography '("~/Workspace/Engineering-Thesis/paper/Bibliography.bib"))
(setq org-latex-pdf-process (list "latexmk -shell-escape -bibtex -f -pdf %f"))
